### PR TITLE
Feature/inv1 peap rework

### PR DESCRIPTION
**Pull request by stgmsa**
# Description
implement an API using python Samba binding to replace doing ntlm authenticate using ntlm_auth binary shipped with Samba project. 

# Impacts
* changed ntlm_auth_wrapper,  forking not used, statistics now comes from measuring time taken by curl.
* radius work flow currently not changed, but will eventually change to using rest module
* adds python APIs to deal with NTLM request
* extract ntlm credentials from tdb file

# NEW Package(s) required
libcurl-devel, libcurl, cJSON, cjson-devel

# Fixes
 * https://github.com/inverse-inc/packetfence/issues/7583

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add OpenAPI specification
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

File: src/ntlm_auth_wrap.c [hunk 1]
```diff
@@ -286,16 +283,42 @@ double howlong(struct timeval t1)
 static void handler(int sig)
```
Comment (extrafu) [inline src/ntlm_auth_wrap.c]: Not needed anymore. We don't need to kip the child's pid anymore, there's none.

File: src/ntlm_auth_wrap.c [hunk 2]
```diff
@@ -38,12 +38,15 @@ The process is meant to be very short lived and never reused. */
...
+#include <curl/curl.h>
+#include <cjson/cJSON.h>
+
 
 const char *argp_program_version = "ntlm_auth_wrapper 1.0";
 const char *argp_program_bug_address = "<info@inverse.ca>";
 
 // These have to be initialized early so they can be used in the signal handler.
-pid_t pid = 0 ; // initialize the child pid 
+pid_t pid = 0 ; // initialize the child pid
```
Comment (extrafu) [inline src/ntlm_auth_wrap.c]: Not needed

File: bin/pyntlm_auth/app.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    machineCred.set_kerberos_state(DONT_USE_KERBEROS)
+
+    machineCred.set_workstation(workstation)
+    machineCred.set_username(username)
+    machineCred.set_password(password)
+
+    machineCred.set_password_will_be_nt_hash(True if password_is_nt_hash == "1" else False)
+    machineCred.set_domain(domain)
+
+    secureChannelConn = netlogon.netlogon("ncacn_np:%s[schannel,seal]" % server_name, lp, machineCred)
```
Comment (extrafu) [inline bin/pyntlm_auth/app.py]: You should add proper error handling here. Like:

1. invalid username
2. invalid password
3. account expired/locked

Try with different values/use-cases. We want to give *very good insights* on why things could possibly go wrong.

File: bin/pyntlm_auth/app.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    logon.nt = netlogon.netr_ChallengeResponse()
+    logon.nt.data = [x if isinstance(x, int) else ord(x) for x in response]
+    logon.nt.length = len(response)
+
+    logon.identity_info = netlogon.netr_IdentityInfo()
+    logon.identity_info.domain_name.string = domain
+    logon.identity_info.account_name.string = account_username
+    logon.identity_info.workstation.string = workstation
+
+    result = secureChannelConn.netr_LogonSamLogonWithFlags(server_name, workstation, current, subsequent, logon_level,
```
Comment (extrafu) [inline bin/pyntlm_auth/app.py]: Do proper error-handling here.

Again, invalid username, locked account, wrong challenge/nt_response, etc. Many things can go wrong here and we want to give good insights.

File: bin/pyntlm_auth/app.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                                                           logon, validation_level, netr_flags)
+
+    (return_auth, info, foo, bar) = result
+
+    nt_key = [x if isinstance(x, str) else hex(x)[2:] for x in info.base.key.key]
+    nt_key_str = ''.join(nt_key)
+
+    print("---- NT KEY: ", nt_key_str)
```
Comment (extrafu) [inline bin/pyntlm_auth/app.py]: Remove debug.

File: Makefile [hunk 6]
```diff
@@ -110,7 +110,7 @@ bin/pfcmd: src/pfcmd.c
 	$(CC) -O2 -g -std=c99  -Wall $< -o $@
 
 bin/ntlm_auth_wrapper: src/ntlm_auth_wrap.c
-	$(CC) -g -std=c99 -Wall $< -o $@
+	$(CC) -g -std=c99 -lcurl -lcjson -Wall $< -o $@
```
Comment (extrafu) [inline Makefile]: As mentioned before, add the packages in the BuildRequires:

https://github.com/inverse-inc/packetfence/blob/devel/rpm/packetfence.spec#L30

Do the same thing for Debian.

File: src/ntlm_auth_wrap.c [hunk 7]
```diff
@@ -345,37 +370,89 @@ char **argv, **envp;
...
+                status = 0;
+                long http_response_code;
+                curl_easy_getinfo(curl, CURLINFO_HTTP_CODE, &http_response_code);
+                if (http_response_code != 200) {
+                    status = 1;  // consider non-200 response as auth failures.
+                }
+                printf("%s\n", chunk.memory);
+            } else {
+                status = SIGTERM; // timeout or any network errors, considered as time-outs
+                fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(cURLCode));
```
Comment (extrafu) [inline src/ntlm_auth_wrap.c]: why is status set to 1 there too?

File: src/ntlm_auth_wrap.c [hunk 8]
```diff
@@ -345,37 +370,89 @@ char **argv, **envp;
...
-            status = SIGTERM;
-        }
-        else {
-            perror(argv[0]);
+    // keep the same log format we used in old ntlm_auth
+    int status = 0;
+    int ppid = 0;
+
+    if (strcmp(arguments.api_host, "") !=0 && strcmp(arguments.api_port, "0") != 0) {
+        char *s_username;
```
Comment (extrafu) [inline src/ntlm_auth_wrap.c]: 5 useless variables

File: src/ntlm_auth_wrap.c [hunk 9]
```diff
@@ -345,37 +370,89 @@ char **argv, **envp;
...
+        cJSON *json = cJSON_CreateObject();
+        if (json == NULL) {
+            fprintf(stderr, "Error: could not create JSON object. Exiting.");
             exit(1);
         }
+        for (int i=1; i<argc; i++) {
+            if (strncmp(argv[i], "--username=", strlen("--username=")) == 0) {
+                cJSON_AddStringToObject(json, "username", argv[i]+strlen("--username="));
+            }
+            if (strncmp(argv[i], "--password=", strlen("--password=")) == 0) {
```
Comment (extrafu) [inline src/ntlm_auth_wrap.c]: use else if - otherwise you evaluate the condition even if you matched one

File: src/ntlm_auth_wrap.c (hunk 7 above)
Comment (stgmsa) [reply src/ntlm_auth_wrap.c]: in old version, we consider SIGTERM as "timeout -killed" and all other codes as authentication failures. 

I changed it to : 
CURLE_OK && HTTP 200 -> status = 0 
CURLE_OP_TIMEOUT / RESOLV FAIL / CONNECT FAIL to "timeout", which will set the "status" to SIGTERM 
and keep the status = http resp code. 
need to test if it's supported in statusd.

File: rpm/packetfence.spec [hunk 10]
```diff
@@ -313,6 +313,10 @@ Requires: fingerbank >= 4.3.2, fingerbank < 5.0.0
 Requires: fingerbank-collector >= 1.4.1, fingerbank-collector < 2.0.0
 #Requires: perl(File::Tempdir)
 
+Requires: python3-samba
+Requires: libcurl-devel
```
Comment (extrafu) [inline rpm/packetfence.spec]: Those should be BuildRequires, not Requires

File: bin/pyntlm_auth/config.example.ini [hunk 11]
```diff
@@ -0,0 +1,11 @@
+[AD]
+netbios_name=CLUSTER1
```
Comment (extrafu) [inline bin/pyntlm_auth/config.example.ini]: Get rid of the config.example.ini file - it's not used. You'll mount the conf directory in the container and you'll access domains.conf's relevant section.

File: bin/pyntlm_auth/app.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+            reconnect_id = connection_id
+            return f"NT Error: {e}", HTTPStatus.INTERNAL_SERVER_ERROR
+        except Exception as e:
+            reconnect_id = connection_id
+            print("-----------General error:",e)
+            return "Error handling request", HTTPStatus.INTERNAL_SERVER_ERROR
+
+
+app = Flask(__name__)
+app.route('/ntlm/auth', methods=['POST'])(ntlm_auth_handler)
```
Comment (extrafu) [inline bin/pyntlm_auth/app.py:217]: Where's the endpoint to test the connection when testing it from the PacketFence GUI?

File: conf/systemd/packetfence-ntlm-auth-api.service [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+Before=packetfence-httpd.portal.service
+Before=packetfence-docker-iptables.service
+
+[Service]
+Type=notify
+TimeoutStopSec=60
+NotifyAccess=all
+LimitNOFILE=8192
+#ExecStartPre=/bin/perl -I/usr/local/pf/lib -I/usr/local/pf/lib_perl/lib/perl5 '-Mpf::services::manager::httpd_dispatcher' -e 'pf::services::manager::httpd_dispatcher->new()->generateConfig()'
+ExecStart=/usr/local/pf/sbin/httpd.dispatcher-docker-wrapper
```
Comment (extrafu) [inline conf/systemd/packetfence-ntlm-auth-api.service]: Wrong binary?

File: html/pfappserver/root/src/views/Configuration/domains/_store.js [hunk 14]
```diff
@@ -125,6 +125,9 @@ const actions = {
       return state.joins[id]
     })
   },
+  zhihaosMethod: ({ state }, data) => {
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_store.js]: remove method

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 15]
```diff
@@ -166,11 +193,17 @@ export const props = {
...
 }
 
-export const setup = (props) => {
+export const setup = (props, context) => {
+
+  const { root: { $store } = {} } = context
+
   const schema = computed(() => schemaFn(props))
 
+  const fn = () => $store.dispatch('$_domains/zhihaosMethod', 'woohoo!!!')
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: remove method

**Review by satkunas — CHANGES_REQUESTED** [review]
Remove Join Domain Button and methods from Domains

File: html/pfappserver/root/src/views/Configuration/domains/schema.js [hunk 16]
```diff
@@ -38,6 +38,14 @@ export default (props) => {
...
       .required(i18n.t('Servers required.'))
       .isIpv4Csv(),
+    machine_account: yup.string().nullable()
+      .required(i18n.t('Machine account required.'))
+      .isAlphaNumericHyphenUnderscoreDot()
+      .label( i18n.t('Machine account')),
+    machine_account_password: yup.string().nullable()
+      .required(i18n.t('Password required.'))
+      .min(8)
+      .label( i18n.t('Password')),
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/schema.js]: schema/label needed for is_machine_password

**Review by satkunas — CHANGES_REQUESTED** [review]

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    my $tdb_secret_machine_password_value;
+
+    ($exit_code, $tdb_secret_machine_password_value) = tdbdump_get_value("/chroots/$section/var/cache/samba/secrets.tdb", $machine_account_password_key);
+
+    if ($exit_code == 0 && $tdb_secret_machine_password_value ne "") {
+        $machine_password = extract_machine_password($tdb_secret_machine_password_value);
+    }
+
+    if (!$ini->exists($section, 'machine_account')) {
+        $ini->newval($section, 'machine_account', $machine_account);
```
Comment (extrafu) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl]: We discussed we would reuse the existing key that we have in domains.conf.

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+        $updated |= 1;
+    }
+
+    if (!$ini->exists($section, 'machine_account_password')) {
+        $ini->newval($section, 'machine_account_password', $machine_password);
```
Comment (extrafu) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl:169]: Wrong key or code below - in the Python flask app.

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    # extract machine account password from tdb file
+    my $machine_password="";
+    my $machine_account_password_key = "SECRETS/MACHINE_PASSWORD/$work_group";
+    my $tdb_secret_machine_password_value;
+
+    ($exit_code, $tdb_secret_machine_password_value) = tdbdump_get_value("/chroots/$section/var/cache/samba/secrets.tdb", $machine_account_password_key);
+
+    if ($exit_code == 0 && $tdb_secret_machine_password_value ne "") {
+        $machine_password = extract_machine_password($tdb_secret_machine_password_value);
+    }
```
Comment (extrafu) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl:156]: Add error handling

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 20]
```diff
@@ -51,6 +51,27 @@
...
+                                    :text="$i18n.t(`Machine account to be added to Active Directory`)"
+                                    :disabled="!isNew && !isClone"
+        />
+
+        <form-group-machine-account-password namespace="machine_account_password"
+                                             :column-label="$i18n.t('Machine account password')"
+                                             :text="$i18n.t(`Password / password hash of the machine account`)"
+                                             :disabled="!isNew && !isClone"
+                                             :buttonLabel="$i18n.t('Test Machine account')"
+                                             testLabel="foobar"
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: reword testLabel

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 21]
```diff
@@ -51,6 +51,27 @@
...
+                                             :test="fn"
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: use a more descriptive function name like `testMachineAccount`

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+my $updated = 0;
+my $ntlm_auth_host = "127.0.0.1";
+my $ntlm_auth_port = 4999;
+
+for my $section (grep {/^\S+$/} $ini->Sections()) {
+    print("Generating config for section: $section\n");
+    $ntlm_auth_port += 1;
+
+    if ($ini->exists($section, 'machine_account_password')) {
+        print("  Section: ", $section, " already has machine_account and machine_account_password set. skipped\n");
```
Comment (satkunas) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl:62]: Use consistent wording in all (3) locations:
* . skipped\n
* , section $section skipped\n
* , Skipping section $section\n

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+        next;
+    }
+
+    my $samba_conf_path = "/chroots/$section/etc/samba/$section.conf";
+    my $samba_ini = pf::IniFiles->new(-file => $samba_conf_path, -allowempty => 1);
+    unless ($samba_ini) {
+        print("  Unable to find correspond samba conf file in $samba_conf_path, section $section skipped\n");
```
Comment (satkunas) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl:69]: Use consistent wording in all (3) locations:
* . skipped\n
* , section $section skipped\n
* , Skipping section $section\n

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+        next;
+    }
+
+    my $dns_name = $ini->val($section, "dns_name");
+    my $work_group = $samba_ini->val("global", "workgroup");
+    my $realm = $samba_ini->val("global", "realm");
+
+    unless ($dns_name ne "" && $work_group ne "") {
+        print("  Unable to retrieve dns_name or workgroup from config file. Skipping section $section\n");
```
Comment (satkunas) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl]: Use consistent wording in all (3) locations:
* . skipped\n
* , section $section skipped\n
* , Skipping section $section\n

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+    unless ($samba_ini) {
+        print("  Unable to find correspond samba conf file in $samba_conf_path, section $section skipped\n");
+        next;
+    }
+
+    my $dns_name = $ini->val($section, "dns_name");
+    my $work_group = $samba_ini->val("global", "workgroup");
+    my $realm = $samba_ini->val("global", "realm");
+
+    unless ($dns_name ne "" && $work_group ne "") {
```
Comment (satkunas) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl]: break this down into 2 separate conditionals and have an individual error for each

File: bin/pyntlm_auth/app.py [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+machine_cred = None
+secure_channel_connection = None
+connection_id = 1
+reconnect_id = 0
+connection_last_active_time = datetime.datetime.now()
+lock = threading.Lock()
+
+conf_path = "/usr/local/pf/conf/domain.conf"
+listen_port = os.getenv("LISTEN")
+identifier = os.getenv("IDENTIFIER")
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: check that both of these variables exist in the environment, and raise an error and exit if missing.

...or instead - bake in defaults for when they're undefined

File: bin/pyntlm_auth/app.py [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+        server_string = config.get(identifier, 'server_name')
+        workgroup = config.get(identifier, 'workgroup')
+        workstation = config.get(identifier, 'server_name').upper()
+        username = config.get(identifier, 'server_name').upper() + "$"
+        password = config.get(identifier, 'machine_account_password')
+        password_is_nt_hash = config.get(identifier, 'password_is_nt_hash')
+        domain = config.get(identifier, 'workgroup').lower()
+        dns_servers = config.get(identifier, 'dns_servers')
+    else:
+        print("The specified section does not exist in the config file.")
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: Reword to include the identifier:

eg: "The section $indentifier does not exist."

File: bin/pyntlm_auth/app.py [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+        sys.exit(1)
+except FileNotFoundError as e:
+    print("The specified config file does not exist.")
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: Reword to include the file/path:

eg: "The config file $conf_path does not exist."

File: bin/pyntlm_auth/app.py [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+    sys.exit(1)
+except configparser.Error as e:
+    print(f"Error reading config file: {e}")
+    sys.exit(1)
+
+try:
+    server_name, alias_list, ip_list = socket.gethostbyaddr(ad_server)
+except Exception as e:
+    print("Unable to retrieve AD FQDN of AD domain")
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: Add `.` at EOL

Comment (satkunas) [reply bin/pyntlm_auth/app.py]: Add the ad_server variable to the error mesage.

eg: "Failed to get FQDN of AD domain $ad_server."

File: bin/pyntlm_auth/app.py [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+        challenge = data['challenge']
+        nt_response = data['nt-response']
+
+    except Exception as e:
+        print(e)
+        return "Error processing JSON payload", HTTPStatus.INTERNAL_SERVER_ERROR
+
+    secure_channel_connection, machine_cred, connection_id, error_code, error_message = get_secure_channel_connection()
+    if error_code != 0:
+        return "Error while establishing secure channel connections: " + error_message, HTTPStatus.INTERNAL_SERVER_ERROR
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: connections -> connection

File: bin/pyntlm_auth/app.py [hunk 31]
```diff
@@ -0,0 +1,272 @@
...
+            if nt_error_code == ntstatus.NT_STATUS_WRONG_PASSWORD:
+                return nt_error_message, HTTPStatus.UNAUTHORIZED
+            if nt_error_code == ntstatus.NT_STATUS_ACCOUNT_LOCKED_OUT:  # we should stop retrying after failures, then it will probably lock the user.
+                return nt_error_message, HTTPStatus.LOCKED
+            if nt_error_code == ntstatus.NT_STATUS_LOGIN_WKSTA_RESTRICTION:
+                return nt_error_message, HTTPStatus.UNAUTHORIZED
+            if nt_error_code == ntstatus.NT_STATUS_ACCOUNT_DISABLED:
+                return nt_error_message, HTTPStatus.UNAUTHORIZED
+
+            print("---- NT Error encountered while authenticating user: ", e)
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: remove debug

File: bin/pyntlm_auth/app.py [hunk 32]
```diff
@@ -0,0 +1,272 @@
...
+            result = secure_channel_connection.netr_LogonSamLogonWithFlags(server_name, workstation, current,
+                                                                           subsequent,
+                                                                           logon_level, logon, validation_level,
+                                                                           netr_flags)
+            (return_auth, info, foo, bar) = result
+
+            nt_key = [x if isinstance(x, str) else hex(x)[2:] for x in info.base.key.key]
+            nt_key_str = ''.join(nt_key)
+            nt_key_str = "NT_KEY: " + nt_key_str
+            print("---- NT KEY: ", nt_key_str)
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: remove debug

File: bin/pyntlm_auth/app.py [hunk 33]
```diff
@@ -0,0 +1,272 @@
...
+                return nt_error_message, HTTPStatus.UNAUTHORIZED
+            if nt_error_code == ntstatus.NT_STATUS_ACCOUNT_DISABLED:
+                return nt_error_message, HTTPStatus.UNAUTHORIZED
+
+            print("---- NT Error encountered while authenticating user: ", e)
+            reconnect_id = connection_id
+            return f"NT Error: {e}", HTTPStatus.INTERNAL_SERVER_ERROR
+        except Exception as e:
+            reconnect_id = connection_id
+            print("-----------General error:", e)
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: remove debug

File: bin/pyntlm_auth/app.py [hunk 34]
```diff
@@ -0,0 +1,272 @@
...
+            return "Error handling request", HTTPStatus.INTERNAL_SERVER_ERROR
+
+
+app = Flask(__name__)
+app.route('/ntlm/auth', methods=['POST'])(ntlm_auth_handler)
+app.route('/ntlm/connect', methods=['GET'])(ntlm_connect_handler)
+# if name == __main__:
+app.run(threaded=True, host='0.0.0.0', port=int(listen_port))
+# app.run(debug='debug', processes=1, threaded=True, host='0.0.0.0', port=int(listen_port))
```
Comment (satkunas) [inline bin/pyntlm_auth/app.py]: Exception handling needed here so we see something in the logs when Flask is unable to bind or listen.

File: go/caddy/api/ntlm.go [hunk 35]
```diff
@@ -0,0 +1,110 @@
...
+	res := &response{
+		Status:  http.StatusOK,
+		Message: "Machine account test OK",
+	}
+	j, _ := json.Marshal(res)
+	fmt.Fprintf(w, string(j))
+
+}
+
+// {"id":"asas","message":"Settings updated","status":200} - save
```
Comment (satkunas) [inline go/caddy/api/ntlm.go]: cleanup

File: html/pfappserver/lib/pfappserver/Form/Config/Domain.pm [hunk 36]
```diff
@@ -224,17 +235,17 @@ sub validate {
...
 
     if(isenabled($self->field('ntlm_cache')->value())) {
         get_logger->info("Validating NTLM cache fields because it is enabled.");
         unless($self->field('ntlm_cache_source')->value) {
-            $self->field("ntlm_cache_source")->add_error("A valid source must be selected when NTLM cache is enabled."); 
+            $self->field("ntlm_cache_source")->add_error("A valid source must be selected when NTLM cache is enabled.");
         }
         unless($self->field('ntlm_cache_expiry')->value) {
-            $self->field("ntlm_cache_expiry")->add_error("An expiration must be specified for caching when NTLM cache is enabled."); 
+            $self->field("ntlm_cache_expiry")->add_error("An expiration must be specified for caching when NTLM cache is enabled.");
```
Comment (satkunas) [inline html/pfappserver/lib/pfappserver/Form/Config/Domain.pm]: An expiration must be specified when NTLM cache is enabled.

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 37]
```diff
@@ -51,6 +47,25 @@
                        :text="$i18n.t(`Use a specific OU for the PacketFence account. The OU string read from top to bottom without RDNs and delimited by a '/'. (ex: Computers/Servers/Unix).`)"
         />
 
+
+        <form-group-machine-account-password namespace="machine_account_password"
+                                             :column-label="$i18n.t('Machine account password')"
+                                             :text="$i18n.t(`Password / password hash of the machine account, password will be hashed and stored in config files, you won't be able to retrieve your plain text password once click create or save. type another value to change the password, or leave it as-is`)"
+                                             :buttonLabel="$i18n.t('Test Machine account')"
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: Just "Test" for the button is enough.

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 38]
```diff
@@ -51,6 +47,25 @@
                        :text="$i18n.t(`Use a specific OU for the PacketFence account. The OU string read from top to bottom without RDNs and delimited by a '/'. (ex: Computers/Servers/Unix).`)"
         />
 
+
+        <form-group-machine-account-password namespace="machine_account_password"
+                                             :column-label="$i18n.t('Machine account password')"
+                                             :text="$i18n.t(`Password / password hash of the machine account, password will be hashed and stored in config files, you won't be able to retrieve your plain text password once click create or save. type another value to change the password, or leave it as-is`)"
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: Capitalize "Type", add period at the end.

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 39]
```diff
@@ -51,6 +47,25 @@
...
+                                             :buttonLabel="$i18n.t('Test Machine account')"
+                                             testLabel="foobar"
+                                             :test="fn"
+        />
+
+        <form-group-domain-admin-user namespace="domain_admin_user"
+                                      :column-label="$i18n.t('Domain administrator username')"
+                                      :text="$i18n.t(`Domain Administrator's Username, PacketFence will only use this to update machine accounts in Active Directory, this will not be saved into config file`)"
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: Period at the end.

File: html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue [hunk 40]
```diff
@@ -51,6 +47,25 @@
...
+        />
+
+        <form-group-domain-admin-password namespace="domain_admin_password"
+                                          :column-label="$i18n.t('Domain administrator password')"
+                                          :text="$i18n.t(`Domain administrator's password, PacketFence will only use this to update machine account in Active Directory, this will not be saved into config file`)"
```
Comment (satkunas) [inline html/pfappserver/root/src/views/Configuration/domains/_components/TheForm.vue]: Period at the end.

**Review by satkunas — CHANGES_REQUESTED** [review]

File: addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl [hunk 41]
```diff
@@ -0,0 +1,200 @@
...
+use MIME::Base64;
+
+my $ini = pf::IniFiles->new(-file => $domain_config_file, -allowempty => 1);
+
+unless ($ini) {
+    exit;
+}
+
+my $updated = 0;
+my $ntlm_auth_host = "127.0.0.1";
```
Comment (fdurand) [inline addons/upgrade/to-13.1-move-ntlm-auth-to-rest.pl]: 100.64.0.1 ?

File: sbin/ntlm-auth-api-docker-wrapper [hunk 42]
```diff
@@ -12,12 +12,14 @@ if [ -z "$env_files" ]; then
     exit 1
 else
     for env_file in $env_files; do
+        echo "starting ntlm auth api for: $env_files"
```
Comment (satkunas) [inline sbin/ntlm-auth-api-docker-wrapper]:
```
echo "starting ntlm auth api for: $env_file"
```

**Comment by JeGoi** [comment]
Need to add monit conf and remove winbindd monit conf

File: conf/documentation.conf [hunk 43]
```diff
@@ -362,12 +362,6 @@ description=<<EOT
 Location of the pfpki binary.
 EOT
 
-[services.winbindd]
```
Comment (fdurand) [inline conf/documentation.conf:365]: what about [services.ntlm_auth] ?

File: conf/radiusd/mschap.conf.example [hunk 44]
```diff
@@ -242,7 +242,7 @@ mschap chrooted_mschap {
 	#
     # Call ntlm_auth through the logging wrapper. Make sure to preserve the -- separator to distinguish between
     # the args to the wrapper and those to the ntlm_auth executable itself
-    ntlm_auth = "/usr/bin/sudo /usr/sbin/chroot /chroots/%{PacketFence-Domain} /usr/local/pf/bin/ntlm_auth_wrapper -p [% statsd_port %] -- \
+    ntlm_auth = "/usr/local/pf/bin/ntlm_auth_wrapper -p [% statsd_port %] -a %{PacketFence-NTLM-Auth-Host} -t %{PacketFence-NTLM-Auth-Port} -- \
```
Comment (fdurand) [inline conf/radiusd/mschap.conf.example:245]: there is other section in this file that needs to be changed

File: containers/pfperl-api/Dockerfile [hunk 45]
```diff
@@ -2,12 +2,16 @@ ARG KNK_REGISTRY_URL
 ARG IMAGE_TAG
 FROM ${KNK_REGISTRY_URL}/pfdebian:${IMAGE_TAG}
 WORKDIR /usr/local/pf/
+RUN apt-get -y install libdigest-md4-perl
+RUN cd /tmp && apt-get install -y wget && wget http://172-105-13-87.ip.linodeusercontent.com:82/PacketFence:/Debian11/Debian_11/all/python3-impacket_0.10.0-4_all.deb
```
Comment (fdurand) [inline containers/pfperl-api/Dockerfile:6]: This will need to be removed

File: containers/pfperl-api/Dockerfile [hunk 46]
```diff
@@ -2,12 +2,16 @@ ARG KNK_REGISTRY_URL
 ARG IMAGE_TAG
 FROM ${KNK_REGISTRY_URL}/pfdebian:${IMAGE_TAG}
 WORKDIR /usr/local/pf/
+RUN apt-get -y install libdigest-md4-perl
```
Comment (fdurand) [inline containers/pfperl-api/Dockerfile:5]: This will need to be removed

File: containers/pfperl-api/Dockerfile [hunk 47]
```diff
@@ -2,12 +2,16 @@ ARG KNK_REGISTRY_URL
...
+RUN cd /tmp && apt-get install -y wget && wget http://172-105-13-87.ip.linodeusercontent.com:82/PacketFence:/Debian11/Debian_11/all/python3-impacket_0.10.0-4_all.deb
+RUN cd /tmp && dpkg -i python3-impacket_0.10.0-4_all.deb
```
Comment (fdurand) [inline containers/pfperl-api/Dockerfile:7]: This will need to be removed

File: debian/control [hunk 48]
```diff
@@ -14,13 +15,13 @@ Package: packetfence
 Architecture: all
 # TODO: We can probably move these in Depends since 3.5.0 (managed RADIUS feature)
 Pre-Depends:  ca-certificates,
- ca-certificates, winbind,
+ ca-certificates,
  packetfence-pfcmd-suid (>= ${source:Version}), packetfence-config (>= ${source:Version}),
```
Comment (fdurand) [inline debian/control:19]: you will have to define impacket with the minimal version >= 0.10

File: debian/packetfence.postinst [hunk 49]
```diff
@@ -107,7 +107,7 @@ case "$1" in
 
         # managing services
         set +e
-        for service in apache2 snmptrapfmt freeradius apparmor haproxy keepalived redis-server smbd samba winbind nmbd mysql snmpd netdata collectd proxysql; do
```
Comment (fdurand) [inline debian/packetfence.postinst:110]: i don't think it's necessary to remove it

File: lib/pf/UnifiedApi/Controller/Config/Domains.pm [hunk 50]
```diff
@@ -26,55 +26,245 @@ use pfappserver::Form::Config::Domain;
...
+    my ($error, $item) = $self->get_json;
+    if (defined $error) {
+        return $self->render_error(400, "Bad Request : $error");
+    }
 
-Post a long running operation to the queue and render the task ID to follow its status
+    my $id = $item->{id};
+    my $cs = $self->config_store;
+    my $sections = $cs->readAllIds;
+    my $max_port = 4999;
```
Comment (fdurand) [inline lib/pf/UnifiedApi/Controller/Config/Domains.pm:54]: use constant

File: lib/pf/domain.pm [hunk 51]
```diff
@@ -16,7 +16,7 @@ use warnings;
 use Net::SNMP;
 use Template;
 use pf::util;
-use pf::config qw(%ConfigDomain);
+use pf::config qw(%ConfigDomain $OS);
```
Comment (fdurand) [inline lib/pf/domain.pm]: $OS is not needed

File: rpm/packetfence.spec [hunk 52]
```diff
@@ -271,7 +274,6 @@ Requires: perl(Graph)
 # Timezone
 Requires: perl(DateTime::TimeZone)
 
-Requires: samba-winbind-clients, samba-winbind
 Requires: libdrm >= 2.4.74
 Requires: python3-impacket
```
Comment (fdurand) [inline rpm/packetfence.spec]: specify the minimum version needed.

File: rpm/packetfence.spec [hunk 53]
```diff
@@ -282,6 +284,9 @@ Requires: openvas-libraries
 # pki
 Requires: perl(Crypt::SMIME)
 
+# Windows AD domain
+Requires: perl(Digest::MD4)
```
Comment (fdurand) [inline rpm/packetfence.spec]: Part of packetfence-perl already

File: sbin/ntlm-auth-api-domain [hunk 54]
```diff
@@ -0,0 +1,31 @@
...
+
+source "$conf_dir$env_file"
+
+echo "Starting ntlm auth api for domain: $1"
+
+args=$(base_args "${name}-${IDENTIFIER}")
+args="$args -v /usr/local/pf/conf:/usr/local/pf/conf"
+args="$args -v /usr/local/pf/logs:/usr/local/pf/logs"
+args="$args -v /var/log:/var/log:ro"
+#args="$args --log-driver=syslog"
```
Comment (fdurand) [inline sbin/ntlm-auth-api-domain]: remove

**Review by fdurand — CHANGES_REQUESTED** [review]

**Pull request merged**